Fix for empty items download

Stefano80 opened the pull request:
This is an emergency fix for the problems caused by a mistake by me.

Since at the moment I am not able to fresh install on a VM, I will check whether I can test it on the dev server, or I hope somebody can check this quickly.

Stefano80 commented:
It looks like I broke it, this should fix.

I am the moment not able to start a fresh install on a VM, it somehow does not work on my machine. Can somebody test it locally?

Stefano80 commented:
By the way, when trying to connect to the dev server, I got a rsa key warning. Does anybody have the same problem?

ppigazzini commented:
@Stefano80 PR tested, here the error message with a Windows worker
```
Downloading 2moves_v1.pgn ...

Exception running games:
Traceback (most recent call last):
  File "D:\__test_worker\ori\worker00\\worker.py", line 91, in worker
    run_games(worker_info, password, remote, run, task_id)
  File "D:\__test_worker\ori\worker00\games.py", line 408, in run_games
    if cleanup(book, testing_dir) and os.path.exists(os.path.join(testing_dir, book)):
  File "D:\__test_worker\ori\worker00\games.py", line 103, in cleanup
    if statinfo.st_size == 0:
NameError: global name 'statinfo' is not defined
Task exited
```

Stefano80 commented:
In have been sloppy sorry. I'll fix and unit test in the evening.

ppigazzini commented:
@Stefano80 sloppiness also by my side.
Lessons learned:
* Test the worker in all possible states (e.g. clean install) before approving for merging
* After merging a worker PR this will be used in a fresh worker installation
* Commit all the worker PRs in a single merge AND change the version number. Otherwise we could have workers in production running different code

ppigazzini commented:
BTW on Ubuntu I use [lxc/lxd](https://stgraber.org/2016/03/11/lxd-2-0-blog-post-series-012/) to build VMs on the fly and test the SW.

Stefano80 commented:
Hi, I now wrote an extensive test and I am quite confident that it should work. @ppigazzini, @tomtor could you give a final check?

ppigazzini commented:
@Stefano80 fresh install of the dev-server and workers with this PR. The workers starts fine with a clean folder.

ppigazzini commented:
@Stefano80 Is this a stand alone PR?

Stefano80 commented:
From my side it is done and could be merged.

ppigazzini commented:
@Stefano80 @tomtor I'm viewing a strange (to me) behavior on the dev server. Yesterday I started three test runs (SPSA, SPRT, Fixed). The last batch of games of the Fixed run was played at 2018-04-03 23:56:43, the others runs (SPRT, SPSA) was finished at 2018-04-04 00:13:05.
Today, at 2018-04-04 06:56:28 CEST, the Fixed run restarted and now is running, now over 80.000 games played, the Actions menu does not show any activity.
Purge could be a factor, but this does not explain the 7 hours gap.
Any thought?

http://94.198.98.239/tests/view/5abeb301e44b0629fedceeb9

Stefano80 commented:
I somehow have the feeling that the server is unstable if there are idle workers. Could this be tested on the dev server?

Stefano80 commented:
And I also noticed that these instabilities happen a lot at around 7 AM CET. Maybe we should start a log of those such that we can find patterns.

ppigazzini commented:
I think the to have a clue about the strange behavior w/ the dev server.

On the dev server, the fishtest service is restarted every 6 hours. For some unknown reasons the service was not able to restart normally at 00:20, the restart at 06:20 solved the problem.

The Fixed run now results purged and pending, with only 17.000 games played.

We should have a log of some events as mongoDB update, run purged etc., the file `~/fishtest/fishtest/nohup.out` is always empty, both on official and dev server.

On the dev server I use systemd, so I have some basic info logged

```
sudo journalctl -u fishtest.service

Apr 03 12:20:01 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 03 12:20:01 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 03 12:20:01 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 03 18:20:01 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 03 18:20:01 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 03 18:20:01 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:15:23 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 04 00:15:23 dfts-0 systemd[1]: Stopped Fishtest Server: connect to port 6543.
Apr 04 00:15:31 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:15:31 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:15:32 dfts-0 systemd[1]: fishtest.service: main process exited, code=exited, status=2/INVALI
Apr 04 00:15:32 dfts-0 systemd[1]: Unit fishtest.service entered failed state.
Apr 04 00:15:32 dfts-0 systemd[1]: fishtest.service failed.
Apr 04 00:15:35 dfts-0 systemd[1]: fishtest.service holdoff time over, scheduling restart.
Apr 04 00:15:35 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:15:35 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.

......

Apr 04 00:19:04 dfts-0 systemd[1]: fishtest.service holdoff time over, scheduling restart.
Apr 04 00:19:04 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:19:04 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:19:05 dfts-0 systemd[1]: fishtest.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Apr 04 00:19:05 dfts-0 systemd[1]: Unit fishtest.service entered failed state.
Apr 04 00:19:05 dfts-0 systemd[1]: fishtest.service failed.
Apr 04 00:19:08 dfts-0 systemd[1]: fishtest.service holdoff time over, scheduling restart.
Apr 04 00:19:08 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:19:08 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:19:09 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 04 00:19:09 dfts-0 systemd[1]: Stopped Fishtest Server: connect to port 6543.
Apr 04 00:19:19 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:19:19 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:20:01 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 04 00:20:01 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:20:01 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 00:30:43 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 04 00:30:43 dfts-0 systemd[1]: Stopped Fishtest Server: connect to port 6543.
Apr 04 00:31:36 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 00:31:36 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 06:20:01 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
Apr 04 06:20:01 dfts-0 systemd[1]: Starting Fishtest Server: connect to port 6543...
Apr 04 06:20:01 dfts-0 systemd[1]: Started Fishtest Server: connect to port 6543.
Apr 04 12:20:01 dfts-0 systemd[1]: Stopping Fishtest Server: connect to port 6543...
```

ppigazzini commented:
@Stefano80 tonight I will retest the PR and, if all OK as expected, i will merge it.

Stefano80 commented on worker/games.py:
```diff
@@ -93,12 +93,30 @@ def verify_signature(engine, signature, remote, payload, concurrency):
+        blob_json = blob_request.json()
+        if blob_request.status_code == requests.codes.ok:
+          break
       blob_json = requests.get(blob['url'], timeout=HTTP_TIMEOUT).json()
```
Thx @tomtor , indeed a piece of the logics here was missing (I was indeed not calling cleanup anywhere). Now it should be fine. Do you have any idea to test this? I should somehow send a request to github and having it to erroneously send an empty file.

Stefano80 commented:
Now after the change the travis test is failing (although it passes locally). It looks like there is suddenly some new problem in communicating to github. It looks like `tree` is not found when downloading.

If I let the worker work locally and let him print the results of the request, everything is fine. But if I execute everything manually (i.e. import requests and then check what is in tree), I get an error. I will keep you updated as I discover what is going on.

ppigazzini commented:
@Stefano80 I manually restarted the Travis build and now all the check have passed.

Stefano80 commented:
Ok, thx I also suspected it was some problem there. Did your nightly test work out well?

ppigazzini commented:
@Stefano80 Tonight I left the dev-server with the https://github.com/glinscott/fishtest/pull/217 to be tested by maintainers/approvers.
I just started the test of this PR.

ppigazzini commented:
@Stefano80 @tomtor to test the PRs I'm using git to update the worker on linux, native windows (msys2), windows subsystem for linux.
I'm wondering if we could use git to push worker updates as python script, books, cutechess-cli etc.
In that case we should run the windows worker using only the msys2 packages (git, python, mingw-w64, make etc.) but this is already possible.

Stefano80 commented:
You mean get rid of the download of blobs using http requests? This sounds like a great idea.

Stefano80 commented:
For your comfort: rebased and squashed.

tomtor commented:
@ppigazzini regarding git to download resources, I think that is a good idea in general for python scripts, etc, but if you use https://github.com/cutechess/cutechess/releases for cutechess then you'll still need to download a file with HTTP?

ppigazzini commented:
@tomtor cutechess official builds are not so up to date and, IIRC, for windows are required some dlls. For Multi Variant Stockfish I routinely build up to date cutechess-cli binaries (w/ QT5 static) using [these scripts](https://github.com/glinscott/fishtest/wiki/Build-cutechess-with-Qt5-static).
At the moment fishtest downloads books and cutechess-cli binaries from the Marco's [Fishcooking](https://github.com/mcostalba/FishCooking/tree/setup) repo.

Stefano80 commented:
@tomtor : I squashed all my changes and by that I cannot have access to your review anymore. Is there anything open from your side?  I thought that I have addressed, but I want to give it a last check. If you say ok, I will merge all the worker stuff.

tomtor commented on worker/games.py:
```diff
@@ -93,13 +93,31 @@ def verify_signature(engine, signature, remote, payload, concurrency):
+        blob_request = requests.get(blob['url'], timeout=HTTP_TIMEOUT)
+        blob_json = blob_request.json()
+        if blob_request.status_code == requests.codes.ok:
+          break
```
I suggest eg a sleep(5) in the loop after the break for if the request.get() fails fast

Stefano80 replied:
Added.

tomtor commented on worker/games.py:
```diff
@@ -93,13 +93,31 @@ def verify_signature(engine, signature, remote, payload, concurrency):
+  if os.path.exists(target_path):
+    stat_info = os.stat(target_path)
+  else:
+    return True
```
Return true when the file does not exist

tomtor commented on worker/games.py:
```diff
@@ -93,13 +93,31 @@ def verify_signature(engine, signature, remote, payload, concurrency):
+    return True
+  if stat_info.st_size == 0:
+    os.remove(target_path)
+    return False
```
Return false when the file does no longer exist. I do not understand the logic related to the previous comment. I would expect the same return value, because in both cases the file must be down loaded?

Stefano80 replied:
The logic is the following: if I have a file 0 in the moment in which I call the cleanup function in the `while` clause, I will remove it. Next time when I go back I will have removed it, so I return `True` and can go on and write down the the file.

The problem I realize now: this will not prevent the worker from writing a 0-bytes file, but it will try to fix it next time. I admit that this is suboptimal and somewhat obscure.

Stefano80 commented:
@ppigazzini : in principle yes, but I just realized (see discussion) that this is a cure for the symptoms but not do solve the problem at the root. Do we want to wait or merge as it is?
